auto_invoice_advance/models/sale_order.py

```diff
@@ -9,32 +9,99 @@
+                                      string="Down payment invoice",
+                                      readonly=True,
+                                      help="Auto generated down payment invoice")
+
+    @api.returns('self', lambda value: value.id)
+    def copy(self, default=None):
+        # OVERRIDE
+
+        if default is None:
+            default = {}
+
+        # Do not copy down_payment_id from other sale orders.
+        if self.down_payment_id:
+            default['down_payment_id'] = None
+
+        return super(SaleOrder, self).copy(default)
+
     def _cron_create_advance_invoices(self):
         order_ids = self.search(self._prepare_order_deposit_domain())
         if len(order_ids) == 0:
             return  # Nothing to be processed
         _logger.info(f"Creating {len(order_ids)} advance invoices for sale.orders: {order_ids.ids}")
 
+        # Get the configurations
         percentage_amount = self.env['ir.config_parameter'].sudo().get_param('sale.deposit_default_percentage')
+        confirm_down_payment = self.env['ir.config_parameter'].sudo().get_param('sale.auto_confirm_down_payment')
+        do_send_email = self.env['ir.config_parameter'].sudo().get_param('sale.auto_email_down_payment')
+
         down_payments_to_create = []
         for order_id in order_ids:
             down_payments_to_create.append(self._prepare_down_payment_values(order_id, percentage_amount))
-        self._create_advance_invoices(down_payments_to_create)
+        self._create_advance_invoices(down_payments_to_create,
+                                      confirm_down_payment,
+                                      do_send_email)
 
-    def _create_advance_invoices(self, down_payments_to_create):
+    def _create_advance_invoices(self, down_payments_to_create, confirm_down_payment, do_send_email):
         payment_inv_ids = self.env['sale.advance.payment.inv'].create(down_payments_to_create)
+        invoice_confirmable_ids = self.env['account.move']
         for payment_inv_id in payment_inv_ids:
-            payment_inv_id.create_invoices()  # singleton check in code, so unable to bulk process
-            self.env.cr.commit()
+            response = payment_inv_id.create_down_payment_invoices()
+            invoice_id = response.get('invoice_id')
+
+            if response.get('can_be_confirmed'):
+                invoice_confirmable_ids += invoice_id
+
+            # Link the created down payment invoice to the sale order(s)
+            # Note that there can be multiple sale orders on a single invoice, link all of them to the invoice
+            for sale_order_id in payment_inv_id.sale_order_ids:
+                sale_order_id.down_payment_id = invoice_id
+
+        if confirm_down_payment:
+            invoice_confirmable_ids.action_post()
+            if do_send_email:
+                invoice_confirmable_ids.print_and_send_automatic_emails()
+
+        # Done with our batch, keep it
+        self.env.cr.commit()
```

Comment:
Why would you do this here in the end? It adds no value as the end of the function would be auto-commited either way.
Instead your `commit()` should go off right after every commit/do_send_email run.
This case if you create, confirm and email an invoice and it crashes somewhere in there you cannot have a cron job that keeps generating/looping every x time. By doing a `commit()` after an important operation you are sure everything is saved and it cannot happen again.